### PR TITLE
Adding additional leading view to TableViewCell

### PR DESCRIPTION
**Pull request by sophialee0416**
### Platforms Impacted
- [x] iOS
- [ ] macOS

### Description of changes

Adding a new "unread dot" view to the TableView Cell.

Binary change:
| File | Before | After | Delta |
|------|--------|-------|-------|
| libFluentUI.a | 28,593,248 bytes | 28,617,896 bytes | +24,648 bytes |
| TableViewCell | 818,360 bytes | 835,600 bytes | +17,240 bytes |
| TableViewCellTokenSet | 118,528 bytes | 118,424 bytes | -104 bytes |

### Verification

| Default                                       | In Selection Mode                                      |
|----------------------------------------------|--------------------------------------------|
| image | image |
| image | image |
| image | image |

Testing RTL: 
image

Testing unread/read dot changes in Selection mode:
Simulator Screen Recording - iPhone 14 Pro Max - 2022-12-09 at 17 41 56

Testing theming:
Simulator Screen Recording - iPhone 14 Pro Max - 2022-12-09 at 17 43 37

### Pull request checklist

This PR has considered:
- [x] Light and Dark appearances
- [ ] iOS supported versions (all major versions greater than or equal current target deployment version)
- [ ] VoiceOver and Keyboard Accessibility
- [x] Internationalization and Right to Left layouts
- [ ] Different resolutions (1x, 2x, 3x)
- [ ] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)
- [ ] iPad [Pointer interaction](https://developer.apple.com/documentation/uikit/pointer_interactions)
- [ ] [SwiftUI](https://developer.apple.com/tutorials/swiftui) consumption (validation or new demo scenarios needed)
- [ ] Objective-C exposure (provide it only if needed)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/fluentui-apple/pull/1446)

### CODE REVIEW

File: ios/FluentUI/Table View/TableViewCell.swift [hunk 1]
```diff
@@ -1506,10 +1529,24 @@ open class TableViewCell: UITableViewCell, TokenizedControlInternal {
             )
         }
 
+        if let cellLeadingView = cellLeadingView {
+            let customViewDimensions: CGFloat = tokenSet[.cellLeadingViewDimensions].float
+            let customViewYOffset = ceil((contentView.frame.height - customViewDimensions) / 2)
+            let customViewXOffset = TableViewCell.customViewLeadingOffset(isInSelectionMode: isInSelectionMode,
```
Comment (sophialee0416) [inline ios/FluentUI/Table View/TableViewCell.swift:1535]: So I totally could have made my own calculation here instead of using the `customViewLeadingOffset` static method here, but I chose this route because of code consistency + cleanliness. I do think that it might have compromised the binary size tho, because now tokens are being called + passed in every where else it's used in this file. Any thoughts would be appreciated :)

**Comment by sophialee0416** [comment]
I also made this little property take in any UIView, but if we want to be more strict about this particular property and *only* make it as a dot, I can happily make that change as well.

File: ios/FluentUI/Table View/TableViewCell.swift [hunk 2]
```diff
@@ -882,6 +892,19 @@ open class TableViewCell: UITableViewCell, TokenizedControlInternal {
         }
     }
 
+    /// The custom view on the leading edge of the `customView` UIView. Does not display if layoutType is `.oneLine`.
```
Comment (harrieshin) [inline ios/FluentUI/Table View/TableViewCell.swift:895]: why not in .oneline?

File: ios/FluentUI/Table View/TableViewCell.swift [hunk 3]
```diff
@@ -1506,10 +1529,24 @@ open class TableViewCell: UITableViewCell, TokenizedControlInternal {
             )
         }
 
+        if let cellLeadingView = cellLeadingView {
+            let cellLeadingViewDimensions: CGFloat = tokenSet[.cellLeadingViewDimensions].float
+            let cellLeadingViewYOffset = ceil((contentView.frame.height - cellLeadingViewDimensions) / 2)
+            let cellLeadingViewXOffset = TableViewCell.customViewLeadingOffset(isInSelectionMode: isInSelectionMode,
```
Comment (sophialee0416) [inline ios/FluentUI/Table View/TableViewCell.swift:1535]: So I totally could have made my own calculation here instead of using the customViewLeadingOffset static method here, but I chose this route because of code consistency + cleanliness. I do think that it might have compromised the binary size tho, because now tokens are being called + passed in every where else it's used in this file. Any thoughts would be appreciated :)

File: ios/FluentUI/Table View/TableViewCell.swift (hunk 2 above)
Comment (sophialee0416) [reply ios/FluentUI/Table View/TableViewCell.swift:895]: Specs only want it in multi-line

File: ios/FluentUI/Table View/TableViewCell.swift [hunk 4]
```diff
@@ -882,6 +892,19 @@ open class TableViewCell: UITableViewCell, TokenizedControlInternal {
         }
     }
 
+    /// The custom view on the leading edge of the `customView` UIView. Does not display if layoutType is `.oneLine`.
+    @objc open var cellLeadingView: UIView? {
+        didSet {
+            if layoutType == .oneLine {
```
Comment (harrieshin) [inline ios/FluentUI/Table View/TableViewCell.swift:898]: having this customizable cellleadinView is awesome! but i think we should also havea default circle drawn if it is part of fluent library now?

File: ios/FluentUI/Table View/TableViewCell.swift (hunk 2 above)
Comment (sophialee0416) [reply ios/FluentUI/Table View/TableViewCell.swift:895]: Actually I suppose if I allow flexibility for the UIView, I can allow flexibility in any layoutType?

File: ios/FluentUI/Table View/TableViewCell.swift [hunk 5]
```diff
@@ -409,6 +412,7 @@ open class TableViewCell: UITableViewCell, TokenizedControlInternal {
     ///   - title: The title string
     ///   - subtitle: The subtitle string
     ///   - footer: The footer string
+    ///   - cellLeadingView: The view on the leading edge of the `customView`
```
Comment (harrieshin) [inline ios/FluentUI/Table View/TableViewCell.swift:415]: something about this name is repetitive. you are already in cell class, so calling cell again seem redundant. you don't say "cellTitle" for example. customLeadingView maybe?

File: ios/FluentUI/Table View/TableViewCell.swift (hunk 4 above)
Comment (sophialee0416) [reply ios/FluentUI/Table View/TableViewCell.swift:898]: I was thinking this as well but wanted to leave options open just in case! It's a quick fix, so I'll go ahead and make that change.

Comment (harrieshin) [reply ios/FluentUI/Table View/TableViewCell.swift:898]: follow example of badges how we use CALayer to draw tiny dots.  layers are more efficient than imageview sometimes.

File: ios/FluentUI/Table View/TableViewCell.swift (hunk 5 above)
Comment (sophialee0416) [reply ios/FluentUI/Table View/TableViewCell.swift:415]: That was my original naming actually, but changed it because it felt also redundant with the existing customView. But maybe with making the view only a pre-set dot, I can just go ahead and name it as leadingDotView? (or the like)

File: ios/FluentUI/Table View/TableViewCell.swift (hunk 4 above)
Comment (sophialee0416) [reply ios/FluentUI/Table View/TableViewCell.swift:898]: Wait sorry, should I have a fallback default circle, or circle only and turn this into a Bool? (etc. showCellLeadingView)

File: ios/FluentUI/Table View/TableViewCell.swift (hunk 2 above)
Comment (harrieshin) [reply ios/FluentUI/Table View/TableViewCell.swift:895]: yeah it feels little strange. "you have all the knobs! jk... only certain times"

**Comment by edjamesmsft** [comment]
Nit: could you explain in the PR description a bit more thoroughly what this change is and why?

File: ios/FluentUI/Table View/TableViewCell.swift (hunk 4 above)
Comment (sophialee0416) [reply ios/FluentUI/Table View/TableViewCell.swift:898]: Resolved offline: Passing in a bool and setting circle internally

File: ios/FluentUI/Table View/TableViewCell.swift (hunk 5 above)
Comment (sophialee0416) [reply ios/FluentUI/Table View/TableViewCell.swift:415]: Resolved with prior convo

File: ios/FluentUI/Table View/TableViewCell.swift (hunk 3 above)
Comment (mischreiber) [reply ios/FluentUI/Table View/TableViewCell.swift:1535]: This definitely feels like the right code design/architecture to be using. Is there a specific binary size concern?

Comment (sophialee0416) [reply ios/FluentUI/Table View/TableViewCell.swift:1535]: This code will be changing so resolving this!

File: ios/FluentUI/Table View/TableViewCell.swift [hunk 6]
```diff
@@ -172,7 +172,7 @@ open class TableViewCell: UITableViewCell, TokenizedControlInternal {
     /// The default leading padding in the cell.
     @objc public static let defaultPaddingLeading: CGFloat = {
         let tokenSet = TableViewCellTokenSet(customViewSize: { .default })
-        return tokenSet[.paddingLeading].float
+        return TableViewCellTokenSet.paddingLeading
```
Comment (edjamesmsft) [inline ios/FluentUI/Table View/TableViewCell.swift:175]: If you're referencing the TableViewTokenSet this way now, do you need the above let? Also what is the rationale for this change?

File: ios/FluentUI.Demo/FluentUI.Demo/TableViewSampleData.swift [hunk 7]
```diff
@@ -25,9 +26,25 @@ class TableViewSampleData {
...
+             hasFooter: Bool = false,
+             footerText: String = "",
+             footerLinkText: String = "",
+             hasCustomLinkHandler: Bool = false,
+             hasCustomAccessoryView: Bool = false,
+             hasCustomLeadingView: Bool = false,
+             hasHandler: Bool = false) {
             self.title = title
             self.items = items
+            self.isUnreadDotVisible = isUnreadDotVisible
```
Comment (edjamesmsft) [inline ios/FluentUI.Demo/FluentUI.Demo/TableViewSampleData.swift:47]: Nit: you call it the leadingDot elsewhere, should we be consistent with naming?

File: ios/FluentUI/Table View/TableViewCell.swift [hunk 8]
```diff
@@ -424,6 +428,7 @@ open class TableViewCell: UITableViewCell, TokenizedControlInternal {
     @objc public class func preferredWidth(title: String,
                                            subtitle: String = "",
                                            footer: String = "",
+                                           hasLeadingDot: Bool = false,
```
Comment (edjamesmsft) [inline ios/FluentUI/Table View/TableViewCell.swift:431]: Is this going to be a breaking change?

File: ios/FluentUI/Table View/TableViewCell.swift (hunk 6 above)
Comment (sophialee0416) [reply ios/FluentUI/Table View/TableViewCell.swift:175]: good catch! Also we are moving away from spacing/padding as customizable tokens, hence the change to constant tokens

File: ios/FluentUI/Table View/TableViewCell.swift [hunk 9]
```diff
@@ -645,16 +652,17 @@ open class TableViewCell: UITableViewCell, TokenizedControlInternal {
     }
 
     private static func customViewLeadingOffset(isInSelectionMode: Bool,
-                                                tokenSet: TableViewCellTokenSet) -> CGFloat {
-        return tokenSet[.paddingLeading].float + selectionModeAreaWidth(isInSelectionMode: isInSelectionMode,
-                                                                        selectionImageMarginTrailing: TableViewCellTokenSet.selectionImageMarginTrailing,
-                                                                        selectionImageSize: TableViewCellTokenSet.selectionImageSize)
+                                                leadingPadding: CGFloat) -> CGFloat {
+        return leadingPadding + selectionModeAreaWidth(isInSelectionMode: isInSelectionMode,
```
Comment (edjamesmsft) [inline ios/FluentUI/Table View/TableViewCell.swift:656]: I like NOT having to pass in the entire token set for a single value. Good change.

File: ios/FluentUI/Table View/TableViewCell.swift [hunk 10]
```diff
@@ -882,6 +890,33 @@ open class TableViewCell: UITableViewCell, TokenizedControlInternal {
         }
     }
 
+    private lazy var unreadDotLayer: CALayer = {
+        let unreadDotLayer = CALayer()
+        let unreadDotSize = TableViewCellTokenSet.leadingDotDimensions
```
Comment (edjamesmsft) [inline ios/FluentUI/Table View/TableViewCell.swift:895]: Why are you referencing TableViewTokenSet directly here, rather than tokenSet?

Comment (sophialee0416) [reply ios/FluentUI/Table View/TableViewCell.swift:894]: Any time I use TableViewTokenSet, that means I'm using a token constant! Which is what leadingDotDimensions happens to be.

File: ios/FluentUI/Table View/TableViewCell.swift (hunk 9 above)
Comment (sophialee0416) [reply ios/FluentUI/Table View/TableViewCell.swift:655]: It was necessary before we had token constants :') Luckily no longer!

File: ios/FluentUI/Table View/TableViewCell.swift (hunk 8 above)
Comment (sophialee0416) [reply ios/FluentUI/Table View/TableViewCell.swift:430]: Good catch, yes it is. I'll put the label on!

File: ios/FluentUI.Demo/FluentUI.Demo/TableViewSampleData.swift (hunk 7 above)
Comment (sophialee0416) [reply ios/FluentUI.Demo/FluentUI.Demo/TableViewSampleData.swift:47]: Yes I totally thought I had replaced everything but clearly not! Changed it!

**Comment by sophialee0416** [comment]
Interestingly, I think changing to CALayer for the unread dot might have increased the binary size further as opposed to using a UIView.

**Comment by edjamesmsft** [comment]
> Interestingly, I think changing to CALayer for the unread dot might have increased the binary size further as opposed to using a UIView.

Do you know by how much? But either way, if the CALayer is more expensive, go for the UIView unless there are clear functional benefits to using the CALayer.

**Comment by sophialee0416** [comment]
> > Interestingly, I think changing to CALayer for the unread dot might have increased the binary size further as opposed to using a UIView.
> 
> Do you know by how much? But either way, if the CALayer is more expensive, go for the UIView unless there are clear functional benefits to using the CALayer.

It was by 4kB I think? It might be because we weren't drawing any views prior. I think sticking to CALayer will be better for now

**Review by edjamesmsft — APPROVED** [review]
